src/storage/aggregator.cc

```diff
@@ -143,48 +177,159 @@ bool Aggregator::Update(const std::string& key, const std::string& row, const ui
+            status_.store(AggrStat::kUnInit, std::memory_order_relaxed);
+            return false;
         }
-        case DataType::kFloat: {
-            float origin_val = *reinterpret_cast<float*>(ch);
-            buffer->aggr_val_.vfloat = origin_val;
+        recovery_offset = std::min(recovery_offset, buffer.binlog_offset_);
+        aggr_latest_offset = std::max(aggr_latest_offset, buffer.binlog_offset_);
+        int64_t latest_ts = buffer.ts_end_ + 1;
+        uint64_t latest_binlog = buffer.binlog_offset_ + 1;
+        buffer.clear();
+        buffer.ts_begin_ = latest_ts;
+        buffer.binlog_offset_ = latest_binlog;
+        if (window_type_ == WindowType::kRowsRange) {
+            buffer.ts_end_ = latest_ts + window_size_ - 1;
+        }
+        it->NextPK();
+    }
+
+    ::openmldb::log::LogReader log_reader(log_parts, base_replicator_->GetLogPath(), false);
+    log_reader.SetOffset(recovery_offset);
+    ::openmldb::api::LogEntry entry;
+    uint64_t cur_offset = recovery_offset;
+    std::string buffer;
+    int last_log_index = log_reader.GetLogIndex();
+    while (true) {
+        buffer.clear();
+        ::openmldb::base::Slice record;
+        ::openmldb::log::Status status = log_reader.ReadNextRecord(&record, &buffer);
+        if (status.IsWaitRecord()) {
+            int end_log_index = log_reader.GetEndLogIndex();
+            int cur_log_index = log_reader.GetLogIndex();
+            if (end_log_index >= 0 && end_log_index > cur_log_index) {
+                log_reader.RollRLogFile();
+                continue;
+            }
             break;
         }
-        case DataType::kDouble: {
-            double origin_val = *reinterpret_cast<double*>(ch);
-            buffer->aggr_val_.vdouble = origin_val;
+        if (status.IsEof()) {
+            if (log_reader.GetLogIndex() != last_log_index) {
+                last_log_index = log_reader.GetLogIndex();
+                continue;
+            }
             break;
         }
-        default: {
-            PDLOG(ERROR, "Unsupported data type");
-            return false;
+
+        if (!status.ok()) {
+            PDLOG(WARNING, "read binlog failed: %s", status.ToString().c_str());
+            continue;
+        }
+
+        bool ok = entry.ParseFromString(record.ToString());
+        if (!ok) {
+            PDLOG(WARNING, "parse binlog failed");
+            continue;
+        }
+        if (cur_offset >= entry.log_index()) {
+            continue;
+        }
+
+        // TODO(nauta): When the base table key is deleted, the pre-aggr table needs to be deleted at the same time.
+        if (entry.has_method_type() && entry.method_type() == ::openmldb::api::MethodType::kDelete) {
+            PDLOG(WARNING, "unsupport delete method for pre-aggr table");
+            continue;
+        }
+        for (int i = 0; i < entry.dimensions_size(); i++) {
+            const auto& dimension = entry.dimensions(i);
```

Comment:
Here I think there is a bug in the aggregation logics (also in [UpdateAggrs](https://github.com/4paradigm/OpenMLDB/blob/47dc3a2cb0007b503eece19f78c134c1c7a2ceed/src/tablet/tablet_impl.cc#L4100)). Each dimension corresponds to one index (i.e., `index_pos` of `CreateAggregatorRequest`).

So different dimension should corresponds to different aggregators. We can only update its related dimension.

In the `TabletImpl::aggregators_`, we should keep the mapping of (tid, pid, index_pos) -> aggregator. Each `dimension` only should be updated in its corresponding aggregator.

Reply:
Double check. current impl should be ok. already do the checking `dimension.idx()` == `index_pos_`